libs/git_wrapper/OGit_wrapper.ml

```diff
@@ -0,0 +1,159 @@
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(* This is like git_wrapper but using pure ocaml https://github.com/mirage/ocaml-git. *)
+(* That way we don't have to shell out *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+
+module Store = Git_unix.Store
+module Hash = Store.Hash
+module Value = Git.Value.Make (Hash)
+module Commit = Git.Commit.Make (Hash)
+module Tree = Git.Tree.Make (Hash)
+module Blob = Git.Blob.Make (Hash)
+module User = Git.User
+
+let tags = Logs_.create_tags [ __MODULE__ ]
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+type object_table = (Value.hash, Value.t) Hashtbl.t
+type hash = Hash.t [@@deriving show, eq, ord]
+type commit = Commit.t [@@deriving show, eq, ord]
+type blob = Blob.t [@@deriving show, eq, ord]
+type author = User.t [@@deriving show, eq, ord]
+
+type blob_with_extra = { blob : blob; path : Fpath.t; size : int }
+[@@deriving show]
+
+(*****************************************************************************)
+(* Reexports *)
+(*****************************************************************************)
+
+let commit_digest = Commit.digest
+let commit_author = Commit.author
+let hex_of_hash = Hash.to_hex
+let blob_digest = Blob.digest
+let string_of_blob = Blob.to_string
+
+(*****************************************************************************)
+(* Code *)
+(*****************************************************************************)
+
+(* A lot of this is from https://github.com/mirage/ocaml-git *)
+let load_store ?(path = Fpath.v (Sys.getcwd ())) () = Store.v path
+
+let load_store_exn ?(path = Fpath.v (Sys.getcwd ())) () =
+  match%lwt load_store ~path () with
+  | Ok store -> Lwt.return store
+  | Error (`Reference_not_found _) ->
+      failwith "Reference not found. Are you in a git repository?"
+  | Error (`Not_found _) -> failwith "Not found. Are you in a git repository?"
+  | Error (`Msg msg) -> failwith msg
+  | _ -> failwith "Unknown error"
+
+let objects_of_store (store : Store.t) : object_table Lwt.t =
+  Logs.debug (fun m -> m ~tags "reading objects");
+  let%lwt contents = Store.contents store in
+  Lwt.return (contents |> List.to_seq |> Hashtbl.of_seq)
+
+let tree_of_commit (objects : object_table) commit =
+  commit |> Commit.tree
+  |> Hashtbl_.find_some objects (fun obj ->
+         match obj with
+         | Git.Value.Tree tree -> tree
+         | _ ->
+             failwith
+               "Not a tree! Shouldn't happen, as we read a tree from a commit \
+                from a store.")
+
+let rec blobs_of_tree ?(path_prefix = "") (objects : object_table)
+    (tree : Tree.t) : blob_with_extra list =
+  tree |> Tree.to_list
+  |> List_.map (blobs_of_entry ~path_prefix objects)
+  |> List.flatten
```

Comment:
`List.concat_map`? (tail-rec too, unlike `List.concat` - flatten is an alias)

Reply:
I always forgot about concat_map